# Segmenting and Clustering Neighborhoods in Downtown Toronto

### Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [37]:
import numpy as np

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)


import json # library to handle JSON files



#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


### Lets' get the joined data from the previous phase (postal codes, boroughs, neigborhoods, latitudes, and longitudes)

In [38]:
Toronto_df = pd.read_csv('Toronto_df.csv')   
Toronto_df.count()

Unnamed: 0      103
PostalCode      103
Borough         103
Neighborhood    103
Latitude        103
Longitude       103
dtype: int64

In [39]:
del Toronto_df['Unnamed: 0']

In [40]:
Toronto_df.head()


PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Port Union,Rouge Hill,Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

### Making sure that the dataset has all 10 boroughs and 103 neighborhoods.

In [41]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(Toronto_df['Borough'].unique()),
       Toronto_df.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


### Using geopy library to get the latitude and longitude values of Toronto

In [42]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


### Creating a map of Toronto with neighborhoods superimposed on top

In [43]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(Toronto_df['Latitude'], Toronto_df['Longitude'], Toronto_df['Borough'], Toronto_df['Neighborhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto


In [44]:
torontoDT_data = Toronto_df[Toronto_df['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
torontoDT_data.head()

PostalCode           Borough                Neighborhood   Latitude  \
0        M4W  Downtown Toronto                    Rosedale  43.679563   
1        M4X  Downtown Toronto  St. James Town,Cabbagetown  43.667967   
2        M4Y  Downtown Toronto        Church and Wellesley  43.665860   
3        M5A  Downtown Toronto                Harbourfront  43.654260   
4        M5B  Downtown Toronto     Garden District,Ryerson  43.657162   

   Longitude  
0 -79.377529  
1 -79.367675  
2 -79.383160  
3 -79.360636  
4 -79.378937

### Let's get the geographical coordinates of Downtown Toronto.

In [45]:
address = 'Downtown Toronto, ON'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6563221, -79.3809161.


### Let's visualize Downtown Toronto the neighborhoods in it.

In [46]:
# create map of Downtown Toronto using latitude and longitude values
map_TorontoDT = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(torontoDT_data['Latitude'], torontoDT_data['Longitude'], torontoDT_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_TorontoDT)  
    
map_TorontoDT

### Defining Foursquare Credentials and Version

In [47]:
CLIENT_ID = '4H0YB13OQY4VY2L12NWC5SGUTLWBQOMUNNGFW45XPMCKEAB0' # your Foursquare ID
CLIENT_SECRET = '4KDXXHI03KW4NIXQPEEQ4JUJOA0EBUM2FOUXETSTKKSIM2W5' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 4H0YB13OQY4VY2L12NWC5SGUTLWBQOMUNNGFW45XPMCKEAB0
CLIENT_SECRET:4KDXXHI03KW4NIXQPEEQ4JUJOA0EBUM2FOUXETSTKKSIM2W5


### Let's explore the first neighborhood in the dataframe.

In [48]:
torontoDT_data.loc[0, 'Neighborhood']

'Rosedale'

In [49]:
neighborhood_latitude = torontoDT_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = torontoDT_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = torontoDT_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rosedale are 43.6795626, -79.37752940000001.


### Let's get the top 100 venues that are in Rosedale within a radius of 500 meters

In [50]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=4H0YB13OQY4VY2L12NWC5SGUTLWBQOMUNNGFW45XPMCKEAB0&client_secret=4KDXXHI03KW4NIXQPEEQ4JUJOA0EBUM2FOUXETSTKKSIM2W5&v=20180605&ll=43.6795626,-79.37752940000001&radius=500&limit=100'

In [51]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e57248caba297001b431721'},
 'response': {'headerLocation': 'Rosedale',
  'headerFullLocation': 'Rosedale, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.6840626045, 'lng': -79.37131878274371},
   'sw': {'lat': 43.675062595499995, 'lng': -79.38374001725632}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4aff2d47f964a520743522e3',
       'name': 'Rosedale Park',
       'location': {'address': '38 Scholfield Ave.',
        'crossStreet': 'at Edgar Ave.',
        'lat': 43.68232820227814,
        'lng': -79.37893434347683,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.68232820227814,
          'lng': -79.37893434347683}],
        'distance': 32

### Let's extract the category of the venue

In [52]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### Let's clean the json and structure it into a pandas dataframe.

In [53]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name  categories        lat        lng
0         Rosedale Park  Playground  43.682328 -79.378934
1          Whitney Park        Park  43.682036 -79.373788
2  Alex Murray Parkette        Park  43.678300 -79.382773
3        Milkman's Lane       Trail  43.676352 -79.373842

### Let's find out how many venues were returned by Foursquare

In [54]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


### Let's repeat the same process to all the neighborhoods in Downtown Toronto

In [55]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Let's create a new dataframe called torontoDT_venues

In [56]:
torontoDT_venues = getNearbyVenues(names= torontoDT_data['Neighborhood'],
                                   latitudes=torontoDT_data['Latitude'],
                                   longitudes=torontoDT_data['Longitude']
                                  )

Rosedale
St. James Town,Cabbagetown
Church and Wellesley
Harbourfront
Garden District,Ryerson
St. James Town
Berczy Park
Central Bay Street
Adelaide,Richmond,King
Harbourfront East,Union Station,Toronto Islands
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
University of Toronto,Harbord
Kensington Market,Chinatown,Grange Park
King and Spadina,Railway Lands,South Niagara,CN Tower,Island airport,Harbourfront West,Bathurst Quay
Stn A PO Boxes 25 The Esplanade
Underground city,First Canadian Place
Christie
Queen's Park


### Let's check the size of the resulting dataframe

In [57]:
print(torontoDT_venues.shape)
torontoDT_venues.head()

(1300, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                    Rosedale              43.679563              -79.377529   
1                    Rosedale              43.679563              -79.377529   
2                    Rosedale              43.679563              -79.377529   
3                    Rosedale              43.679563              -79.377529   
4  St. James Town,Cabbagetown              43.667967              -79.367675   

                    Venue  Venue Latitude  Venue Longitude     Venue Category  
0           Rosedale Park       43.682328       -79.378934         Playground  
1            Whitney Park       43.682036       -79.373788               Park  
2    Alex Murray Parkette       43.678300       -79.382773               Park  
3          Milkman's Lane       43.676352       -79.373842              Trail  
4  Butter Chicken Factory       43.667072       -79.369184  Indian Restaurant

### Let's check how many venues were returned for each neighborhood

In [58]:
torontoDT_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,Richmond,King                                                100   
Berczy Park                                                            55   
Central Bay Street                                                     79   
Christie                                                               18   
Church and Wellesley                                                   84   
Commerce Court,Victoria Hotel                                         100   
Design Exchange,Toronto Dominion Centre                               100   
Garden District,Ryerson                                               100   
Harbourfront                                                           48   
Harbourfront East,Union Station,Toronto Islands                       100   
Kensington Market,Chinatown,Grange Park                                85   
King and Spadina,Railway Lands,South Niagara,CN...                     16   
Queen's Park                                                           40   
Rosedale                                                                4   
St. James Town                                                        100   
St. James Town,Cabbagetown                                             43   
Stn A PO Boxes 25 The Esplanade                                        94   
Underground city,First Canadian Place                                 100   
University of Toronto,Harbord                                          34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,Richmond,King                                                 100   
Berczy Park                                                             55   
Central Bay Street                                                      79   
Christie                                                                18   
Church and Wellesley                                                    84   
Commerce Court,Victoria Hotel                                          100   
Design Exchange,Toronto Dominion Centre                                100   
Garden District,Ryerson                                                100   
Harbourfront                                                            48   
Harbourfront East,Union Station,Toronto Islands                        100   
Kensington Market,Chinatown,Grange Park                                 85   
King and Spadina,Railway Lands,South Niagara,CN...                      16   
Queen's Park                                                            40   
Rosedale                                                                 4   
St. James Town                                                         100   
St. James Town,Cabbagetown                                              43   
Stn A PO Boxes 25 The Esplanade                                         94   
Underground city,First Canadian Place                                  100   
University of Toronto,Harbord                                           34   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Adelaide,Richmond,King                                100             100   
Berczy Park                                            55              55   
Central Bay Street                                     79              79   
Christie                                               18              18   
Church and Wellesley                                   84              84   
Commerce Court,Victoria Hotel                         100             100   
Design Exchange,Toronto Dominion Centre               100             100   
Garden District,Ryerson                               100             100   
Harbourfront             

### Let's find out how many unique categories can be curated from all the returned venues

In [59]:
print('There are {} uniques categories.'.format(len(torontoDT_venues['Venue Category'].unique())))

There are 207 uniques categories.


### Let's analyze each neighborhood in Downtown Toronto

In [60]:
# one hot encoding
torontoDT_onehot = pd.get_dummies(torontoDT_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
torontoDT_onehot['Neighborhood'] = torontoDT_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [torontoDT_onehot.columns[-1]] + list(torontoDT_onehot.columns[:-1])
torontoDT_onehot = torontoDT_onehot[fixed_columns]

torontoDT_onehot.head()


Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                   0          0           0           0       0     0    0   
1                   0          0           0           0       0     0    0   
2                   0          0           0           0       0     0    0   
3                   0          0           0           0       0     0    0   
4                   0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Camera Store  Candy Store  \
0             0              0        0     0             0            0   
1             0              0        0     0             0            0   
2             0              0        0     0             0            0   
3             0              0        0     0             0            0   
4             0              0        0     0             0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               

### Let's examine the new dataframe size.

In [61]:
torontoDT_onehot.shape

(1300, 207)

### Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [62]:
torontoDT_grouped = torontoDT_onehot.groupby('Neighborhood').mean().reset_index()
torontoDT_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,Richmond,King     0.000000   
1                                         Berczy Park     0.000000   
2                                  Central Bay Street     0.012658   
3                                            Christie     0.000000   
4                                Church and Wellesley     0.011905   
5                       Commerce Court,Victoria Hotel     0.000000   
6             Design Exchange,Toronto Dominion Centre     0.000000   
7                             Garden District,Ryerson     0.000000   
8                                        Harbourfront     0.020833   
9     Harbourfront East,Union Station,Toronto Islands     0.000000   
10            Kensington Market,Chinatown,Grange Park     0.000000   
11  King and Spadina,Railway Lands,South Niagara,C...     0.000000   
12                                       Queen's Park     0.025000   
13                                           Rosedale     0.000000   
14                                     St. James Town     0.000000   
15                         St. James Town,Cabbagetown     0.000000   
16                    Stn A PO Boxes 25 The Esplanade     0.000000   
17              Underground city,First Canadian Place     0.000000   
18                      University of Toronto,Harbord     0.029412   

    Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0            0.000000   0.0000              0.0000           0.000   
1            0.000000   0.0000              0.0000           0.000   
2            0.000000   0.0000              0.0000           0.000   
3            0.000000   0.0000              0.0000           0.000   
4            0.011905   0.0000              0.0000           0.000   
5            0.000000   0.0000              0.0000           0.000   
6            0.000000   0.0000              0.0000           0.000   
7            0.000000   0.0000              0.0000           0.000   
8            0.000000   0.0000              0.0000           0.000   
9            0.000000   0.0000              0.0000           0.000   
10           0.000000   0.0000              0.0000           0.000   
11           0.000000   0.0625              0.0625           0.125   
12           0.000000   0.0000              0.0000           0.000   
13           0.000000   0.0000              0.0000           0.000   
14           0.000000   0.0000              0.0000           0.000   
15           0.000000   0.0000              0.0000           0.000   
16           0.000000   0.0000              0.0000           0.000   
17           0.000000   0.0000              0.0000           0.000   
18           0.000000   0.0000              0.0000           0.000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0            0.0000             0.000             0.020000      0.000000   
1            0.0000             0.000             0.000000      0.000000   
2            0.0000             0.000             0.012658      0.000000   
3            0.0000             0.000             0.000000      0.000000   
4            0.0000             0.000             0.011905      0.000000   
5            0.0000             0.000             0.030000      0.000000   
6            0.0000             0.000             0.030000      0.000000   
7            0.0000             0.000             0.010000      0.000000   
8            0.0000             0.000             0.000000      0.020833   
9            0.0000             0.000             0.000000      0.000000   
10           0.0000             0.000             0.000000      0.000000   
11           0.1875             0.125             0.000000      0.000000   
12           0.0000             0.000             0.000000      0.000000   
13           0.0000             0.000             0.000000      0.000000   
14           0.0000             0.000             0.030000      0.000000   
15           0.0000    

### Let's confirm the new size

In [63]:
torontoDT_grouped.shape

(19, 207)

### Let's print each neighborhood along with the top 5 most common venues

In [64]:
num_top_venues = 5

for hood in torontoDT_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = torontoDT_grouped[torontoDT_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,Richmond,King----
             venue  freq
0      Coffee Shop  0.06
1       Restaurant  0.04
2             Café  0.04
3              Bar  0.04
4  Thai Restaurant  0.04


----Berczy Park----
         venue  freq
0  Coffee Shop  0.07
1   Restaurant  0.04
2       Bakery  0.04
3         Café  0.04
4  Cheese Shop  0.04


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.18
1   Italian Restaurant  0.05
2         Burger Joint  0.04
3  Japanese Restaurant  0.04
4            Juice Bar  0.04


----Christie----
                venue  freq
0       Grocery Store  0.22
1                Café  0.17
2                Park  0.11
3          Baby Store  0.06
4  Italian Restaurant  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.10
1  Japanese Restaurant  0.06
2     Sushi Restaurant  0.05
3              Gay Bar  0.05
4           Restaurant  0.04


----Commerce Court,Victoria Hotel----
         venue  freq
0  Coffee Shop 

### Let's put that into a pandas dataframe

In [65]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Let's create the new dataframe and display the top 10 venues for each neighborhood.

In [66]:

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = torontoDT_grouped['Neighborhood']

for ind in np.arange(torontoDT_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(torontoDT_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Adelaide,Richmond,King           Coffee Shop       Thai Restaurant   
1             Berczy Park           Coffee Shop        Farmers Market   
2      Central Bay Street           Coffee Shop    Italian Restaurant   
3                Christie         Grocery Store                  Café   
4    Church and Wellesley           Coffee Shop   Japanese Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Restaurant                   Bar                  Café   
1          Cocktail Bar              Beer Bar                Bakery   
2        Sandwich Place   Japanese Restaurant          Burger Joint   
3                  Park           Candy Store            Restaurant   
4               Gay Bar      Sushi Restaurant            Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Steakhouse        Cosmetics Shop      Sushi Restaurant   
1    Seafood Restaurant           Cheese Shop                  Café   
2             Juice Bar                   Bar           Salad Place   
3            Baby Store    Athletics & Sports                 Diner   
4                  Café           Men's Store             Gastropub   

      9th Most Common Venue 10th Most Common Venue  
0                       Gym              Gastropub  
1                Restaurant                   Park  
2           Bubble Tea Shop                   Café  
3               Gas Station              Nightclub  
4  Mediterranean Restaurant                    Pub

### Let's run k-means to cluster the neighborhood into 3 clusters.

In [67]:
# set number of clusters
kclusters = 3

torontoDT_grouped_clustering = torontoDT_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(torontoDT_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

### Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [68]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

torontoDT_merged = torontoDT_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
torontoDT_merged = torontoDT_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

torontoDT_merged.head() # check the 5 first columns!

PostalCode           Borough                Neighborhood   Latitude  \
0        M4W  Downtown Toronto                    Rosedale  43.679563   
1        M4X  Downtown Toronto  St. James Town,Cabbagetown  43.667967   
2        M4Y  Downtown Toronto        Church and Wellesley  43.665860   
3        M5A  Downtown Toronto                Harbourfront  43.654260   
4        M5B  Downtown Toronto     Garden District,Ryerson  43.657162   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.377529               2                  Park            Playground   
1 -79.367675               0           Coffee Shop    Chinese Restaurant   
2 -79.383160               0           Coffee Shop   Japanese Restaurant   
3 -79.360636               0           Coffee Shop                   Pub   
4 -79.378937               0        Clothing Store           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail         Women's Store         Deli / Bodega   
1            Restaurant                   Pub    Italian Restaurant   
2               Gay Bar      Sushi Restaurant            Restaurant   
3                  Park                Bakery    Mexican Restaurant   
4   Japanese Restaurant                  Café       Bubble Tea Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0   Dumpling Restaurant            Donut Shop      Doner Restaurant   
1                  Café                Bakery           Pizza Place   
2                  Café           Men's Store             Gastropub   
3        Breakfast Spot                  Café               Theater   
4      Ramen Restaurant               Theater                Bakery   

      9th Most Common Venue 10th Most Common Venue  
0                   Dog Run    Distribution Center  
1         Convenience Store         Breakfast Spot  
2  Mediterranean Restaurant                    Pub  
3      Gym / Fitness Center         Farmers Market  
4            Lingerie Store     Italian Restaurant

### Let's visualize the resulting clusters

In [69]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(torontoDT_merged['Latitude'], torontoDT_merged['Longitude'], torontoDT_merged['Neighborhood'], torontoDT_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Cluster 1 

In [70]:
#Cluster 1
torontoDT_merged.loc[torontoDT_merged['Cluster Labels'] == 0, torontoDT_merged.columns[[1] + list(range(5, torontoDT_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0        Clothing Store   
5   Downtown Toronto               0                  Café   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0                  Café   
13  Downtown Toronto               0                   Bar   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0         Grocery Store   
18  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1     Chinese Restaurant            Restaurant                   Pub   
2    Japanese Restaurant               Gay Bar      Sushi Restaurant   
3                    Pub                  Park                Bakery   
4            Coffee Shop   Japanese Restaurant                  Café   
5            Coffee Shop            Restaurant        Clothing Store   
6         Farmers Market          Cocktail Bar              Beer Bar   
7     Italian Restaurant        Sandwich Place   Japanese Restaurant   
8        Thai Restaurant            Restaurant                   Bar   
9               Aquarium                 Hotel                  Café   
10                  Café                 Hotel            Restaurant   
11            Restaurant                  Café                 Hotel   
12                   Bar   Japanese Restaurant             Bookstore   
13                  Café           Coffee Shop                Bakery   
15            Restaurant                  Café                 Hotel   
16                  Café            Restaurant                   Gym   
17                  Café                  Park           Candy Store   
18                  Park                   Gym          Burger Joint   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1      Italian Restaurant                  Café                Bakery   
2              Restaurant                  Café           Men's Store   
3      Mexican Restaurant        Breakfast Spot                  Café   
4         Bubble Tea Shop      Ramen Restaurant               Theater   
5     American Restaurant              Beer Bar        Cosmetics Shop   
6                  Bakery    Seafood Restaurant           Cheese Shop   
7            Burger Joint             Juice Bar                   Bar   
8                    Café            Steakhouse        Cosmetics Shop   
9      Italian Restaurant            Restaurant   Fried Chicken Joint   
10     Italian Restaurant                   Bar             Gastropub   
11                    Gym   American Restaurant    Seafood Restaurant   
12                 Bakery            Restaurant        Sandwich Place   
13  Vietnamese Restaurant    Chinese Restaurant    Mexican Restaurant   
15               Beer Bar   Japanese Restaurant    Italian Restaurant   
16       Asian Restaurant            Steakhouse                 Hotel   
17             Restaurant            Baby Store    Athletics & Sports   
18            Yoga Studio         Burrito Place    Italian Restaurant   

   8th Most Common Venue          9th Most Common Venue 10th Most Common Venue  
1            Pizza Place              Convenience Store         Breakfast Spot  
2              Gastropub       Mediterranean Restaurant                    Pub  
3                Theater           Gym / 

### Cluster 2

In [71]:
#Cluster 2
torontoDT_merged.loc[torontoDT_merged['Cluster Labels'] == 1, torontoDT_merged.columns[[1] + list(range(5, torontoDT_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               1       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14        Airport Lounge      Airport Terminal         Boat or Ferry   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14                 Plane              Boutique   Rental Car Location   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14                   Bar      Sculpture Garden        Harbor / Marina

### Cluster 3

In [72]:
#Cluster 3
torontoDT_merged.loc[torontoDT_merged['Cluster Labels'] == 2, torontoDT_merged.columns[[1] + list(range(5, torontoDT_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0  Downtown Toronto               2                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Playground                 Trail         Women's Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0         Deli / Bodega   Dumpling Restaurant            Donut Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant               Dog Run    Distribution Center

### The discriminating venue category is that Cluster 2 hass only parks, playgrounds, and trails, while the Cluster 1  is mostly all about coffee shops and restaurants. K-means clustering didn't have anything to give in Cluster 3.

### Foursquare's information was not very helpful when dividing Downtown Toronto into clusters using k-means